src/lashup_kv.erl

```diff
@@ -229,72 +233,114 @@ init_db(Nodes) ->
   mnesia:change_table_copy_type (schema, node(), disc_copies), % If the node was already running
   {ok, _} = application:ensure_all_started(mnesia),
   ExistingTables = mnesia:system_info(tables),
-  Tables = [kv],
+  Tables = [?KV_TABLE, nclock],
   TablesToCreate = Tables -- ExistingTables,
+  Alltables = TablesToCreate ++ ExistingTables,
   lists:foreach(fun create_table/1, TablesToCreate),
-  ok = mnesia:wait_for_tables(Tables, infinity).
+  ok = mnesia:wait_for_tables(Alltables, infinity),
+  ok = maybe_upgrade_table(ExistingTables).
 
-create_table(kv) ->
-  {atomic, ok} =  mnesia:create_table(kv, [
-    {attributes, record_info(fields, kv)},
+create_table(Table) ->
+  {atomic, ok} =  mnesia:create_table(Table, [
+    {attributes, get_record_info(Table)},
     {disc_copies, [node()]},
     {type, set}
   ]).
 
+get_record_info(kv2) ->
+  record_info(fields, kv2);
+get_record_info(nclock) ->
+  record_info(fields, nclock).
 
--spec(mk_write_fun(Key :: key(), OldVClock :: riak_dt_vclock:vclock() | undefined, Op :: riak_dt_map:map_op())
-      -> (fun())).
+-spec(maybe_upgrade_table(tables()) -> ok|aborted).
+maybe_upgrade_table(ExistingTables) ->
+  case {lists:member(?KV_TABLE, ExistingTables), lists:member(?OLD_KV_TABLE, ExistingTables)} of
+    {true, _} ->
+        ok; %% already upgraded
+    {false, false} ->
+        ok; %% nothing to upgrade
+    {false, true} ->
+        upgrade_table(kv)
+  end.
+
+-spec(upgrade_table(kv) -> ok|aborted).
+upgrade_table(kv) ->
+  F = fun() ->
+        Records = mnesia:foldl(fun convert_and_write_record/2, [], ?OLD_KV_TABLE),
+        NClock = #nclock{key = node(), lclock = length(Records)},
+        mnesia:write(NClock)
+      end,
+  case mnesia:sync_transaction(F) of
+    {atomic, _} ->
+      ok = mnesia:sync_log();
+    {aborted, Reason} ->
+      lager:error("Failed to upgrade kv table because ~p", [Reason]),
+      aborted
+  end.
+
+convert_and_write_record(Record0, Acc) ->
```

Comment:
Rather than the acc being a list of records shouldn't it just be N? 
Otherwise this algorithm is N**2?